### ISSUE TITLE
Migrating from unique DB for each deployment to 1 DB for all deployments

### ISSUE THREAD

**Issue by jshorland** [ushahidi/platform#1685]
@rjmackay or @tuxpiper can you summarize a description of what we're trying to do here? I think one of you is in a better position than I to scope this properly.

Aha! Link: https://ushahiditeam.aha.io/features/PROD-613

**Comment by tuxpiper** [comment]
I think this means (possibly among other things):

1. inserting a column in all database tables with the deployment id
2. adjusting all the models and queries to take that deployment id into account
3. possibly looking up and setting that deployment id at the beginning of each http request transaction, and background job execution . Default to a special id for non multisite deployments (open source, dedicated servers)
4. modifying the deployment provisioning stuff (when users create deployment) , simplifying it quite a bit actually

@rjmackay may have a more precise idea

**Comment by rjmackay** [comment]
I think a key addition to this is probably: replacing IDs with UUIDs to avoid a clash of primary keys across different deployments.

**Comment by rjmackay** [comment]
- [ ] Add migration to add `deployment_id` column to db tables
- [ ] Adjust Entities and repos to take `deployment_id` into account
- [ ] Replace multisite DB lookup code: currently this switches DBs based on deployment subdomain. I'd suggest we plan for a transition period where new deployments go in a single DB, and we migrate old DBs 1 by 1.
- [ ] Replace deployment creation/provisioning: instead of setting up a new db, just insert a few records in the central DB
- [ ] Create migration scripts ie. for moving single db data into the central DB
    - We might want to convert entity IDs to UUIDs before doing this to avoid primary key clashes
- [ ] Once everything is migrated to 1 db, we can review the code and identify places we can simplify now: ie. job queues, data providers, etc.

**Comment by rjmackay** [comment]
@willdoran I've added a list of what we need to migrate for multi -> single db. Any idea what have I missed ?

**Comment by jshorland** [comment]
@tuxpiper is going to take lead on this. @tuxpiper can you have a draft work plan scoped out by mid-may?

**Comment by tuxpiper** [comment]
Ok, I think Robbie pretty much nailed it :) I will add details around that.

**Comment by tuxpiper** [comment]
Couple comments:
- yesterday we had a nice chat with @rjmackay obbie about how this will impact db migrations, and how careful we've got to be with not breaking deployments people have out there.
- today we chatted with @willdoran about using the database name as a matcher in intercom, with the company/organisation running a deployment in a multisite setup (ushahidi.io) . This will probably require some migration once we move to deployment ids

**Comment by rjmackay** [comment]
@jshorland I don't think this is really near term right now. Should I drop it in backlog? or into the Icebox workspace in a 'this year' column?
